In [1]:
import numpy as np
import matplotlib.pyplot as plt

# Parte I

In [2]:
archiveAll = 'dataAll.txt'
dataAll = np.fromfile(archiveAll)
lenAll = int(len(dataAll)/3)
dataAll = np.reshape(dataAll, (lenAll,3)) 
# Eu sei que é (m,1,3), mas na prática não vimos muita diferença então colocamos (m,3)

dataAll

array([[-363.7884,  244.1423,    0.    ],
       [ 328.7572, -256.7658,    1.    ],
       [-164.9827,  613.2164,    0.    ],
       ...,
       [ 872.4311,  669.1192,    1.    ],
       [ 822.6894, -702.6489,    1.    ],
       [-140.147 ,  734.0983,    0.    ]])

In [3]:
w = np.random.uniform(low=-0.5,high=0.5,size=2)
x = dataAll
err = np.repeat(1, len(x)) #
n = 0.1
aj = 0
print("Pesos iniciais:",w)
epoca = 0
l = []
while (np.isin(1, err)):
    epoca += 1
    print("Época",epoca)
    for i in range(len(x)):
        yd = x[i][2]
        xd = np.array([x[i][0],x[i][1]])
        sigma = w[0]*xd[0]+w[1]*xd[1]
        if (sigma >= 0): # função degrau
            y = 1
        else:
            y = 0
        if (y != yd):
            err[i] = 1
            aj += 1
            w = w + n*(yd-y)*xd
            print(w)
        else:
            err[i] = 0

Pesos iniciais: [0.01899703 0.10907365]
Época 1
[ 36.39783703 -24.30515635]
[ -17.26121297 -100.80090635]
[ 55.53480703 -64.97134635]
[138.13350703  14.88448365]
[143.97926703 -63.41010635]
[ 103.97035703 -141.30279635]
[157.64016703 -92.21052635]
[ 104.68010703 -149.07998635]
[164.23962703 -98.95765635]
[ 118.71499703 -164.59358635]
[209.43085703 -89.88026635]
[ 151.64986703 -150.18841635]
[ 103.38330703 -198.65813635]
[ 173.83065703 -156.54186635]
[ 126.98865703 -207.95173635]
[ 196.85228703 -143.73279635]
[ 141.01655703 -199.91032635]
[ 218.00402703 -135.06099635]
[ 181.73398703 -178.28903635]
[ 163.49484703 -196.62267635]
[ 238.91728703 -130.56756635]
[ 174.55815703 -212.99000635]
[ 265.86636703 -129.71923635]
[ 218.82153703 -217.62317635]
Época 2
[ 170.55497703 -266.09289635]
[ 240.41860703 -201.87395635]
[ 184.58287703 -258.05148635]
[ 261.57034703 -193.20215635]
[ 225.30030703 -236.43019635]
[ 308.19217703 -153.64870635]
[ 262.04479703 -210.29808635]
[ 235.41917703 -242.82045635

In [4]:
print("Foram feitos",aj,"ajustes")
print("E se levaram",epoca,"épocas")

# O número total de ajustes no vetor de pesos;
# (b) O número de épocas até a convergência;
# (c) O gráfico contendo todos os exemplos do conjunto de dados e a reta que separa as classes
# obtida como resultado do treinamento do neurônio Perceptron. Respeitar o esquema de
# cores proposto inicialmente e apresentar a solução de maneira clara neste gráfico.
# x2 = 1.000369759x1

Foram feitos 55 ajustes
E se levaram 4 épocas


# Parte II

In [5]:
# Hiago - 1515280014
# Alex - 1515080008

archive = 'data' + str((4+8) % 4) + '.txt'
data0 = np.fromfile(archive)
len0 = int(len(data0)/3)
data0 = np.reshape(data0, (len0,3))

nl = [0.4,0.1,0.01]
Il = [100, 1, 0.5]
for N in nl:
    for I in Il:
        w = np.random.uniform(low=-I,high=I,size=2)
        x = data0
        err = np.repeat(1, len(x))
        n = N
        aj = 0
        print("Pesos iniciais:",w)
        epoca = 0
        l = []
        while (np.isin(1, err) and epoca < 100):
            epoca += 1
            for i in range(len(x)):
                yd = x[i][2]
                xd = np.array([x[i][0],x[i][1]])
                sigma = w[0]*xd[0]+w[1]*xd[1]
                if (sigma >= 0): # função degrau
                    y = 1
                else:
                    y = 0
                if (y != yd):
                    err[i] = 1
                    aj += 1
                    
                    w = w + n*(yd-y)*xd
                    #print(w)
                else:
                    err[i] = 0
        print("Taxa de aprendizado:",n)
        print("Intervalo:",-I,I)
        print("Pesos finais:",w)
        print("Épocas passadas:",epoca)
        print("Quantidade de certos:",np.count_nonzero(err == 0),"/",len(x))
        print()

Pesos iniciais: [-83.37742123 -96.61759454]
Taxa de aprendizado: 0.4
Intervalo: -100 100
Pesos finais: [0.41688493 0.42883685]
Épocas passadas: 100
Quantidade de certos: 526 / 600

Pesos iniciais: [0.80293458 0.96978233]
Taxa de aprendizado: 0.4
Intervalo: -1 1
Pesos finais: [0.41255325 0.44509258]
Épocas passadas: 100
Quantidade de certos: 525 / 600

Pesos iniciais: [-0.36969668 -0.30868345]
Taxa de aprendizado: 0.4
Intervalo: -0.5 0.5
Pesos finais: [0.42297919 0.42499136]
Épocas passadas: 100
Quantidade de certos: 527 / 600

Pesos iniciais: [78.41348029 55.08419566]
Taxa de aprendizado: 0.1
Intervalo: -100 100
Pesos finais: [-0.00757824  0.0128617 ]
Épocas passadas: 100
Quantidade de certos: 526 / 600

Pesos iniciais: [0.91409927 0.29249616]
Taxa de aprendizado: 0.1
Intervalo: -1 1
Pesos finais: [0.08351078 0.11806696]
Épocas passadas: 100
Quantidade de certos: 527 / 600

Pesos iniciais: [-0.29954241  0.04276265]
Taxa de aprendizado: 0.1
Intervalo: -0.5 0.5
Pesos finais: [0.106535   

# Parte III

In [6]:
archive = 'dataHoldout.txt'
dataH = np.fromfile(archive)
dataH = np.reshape(dataH, (int(len(dataH)/3),3))
dataH

array([[-0.29322959, -0.09065359,  1.        ],
       [ 0.07988839,  0.21101297,  1.        ],
       [-0.07825563, -0.08083512,  1.        ],
       ...,
       [ 0.65980493,  1.05876739,  0.        ],
       [ 1.09867123,  0.87404891,  0.        ],
       [ 0.72930655,  0.93224271,  0.        ]])

In [7]:
# Fazendo o split 70%-30%

np.random.shuffle(dataH)

# Define a size for your train set 
train_size = int(0.7 * len(dataH))

# Split your dataset 
train_set = dataH[:train_size]
test_set = dataH[train_size:]

In [8]:
w = np.random.uniform(low=-1.0,high=1.0,size=2)
x = train_set
err = np.repeat(1, len(x))
n = 0.1
aj = 0
print("Pesos iniciais:",w)
epoca = 0
l = []
while (np.isin(1, err) and epoca < 100):
    epoca += 1
    for i in range(len(x)):
        yd = x[i][2]
        xd = np.array([x[i][0],x[i][1]])
        sigma = w[0]*xd[0]+w[1]*xd[1]
        if (sigma >= 0): # função degrau
            y = 1
        else:
            y = 0
        if (y != yd):
            err[i] = 1
            aj += 1
            
            w = w + n*(yd-y)*xd
            #print(w)
        else:
            err[i] = 0
    np.random.shuffle(train_set)
    # Randomizando o treinamento ao fim de cada época
print("Pesos finais:",w)
print("Quantidade de certos:",np.count_nonzero(err == 0),"/",len(x))

Pesos iniciais: [-0.47159224  0.08074662]
Pesos finais: [-0.08561795 -0.10776963]
Quantidade de certos: 471 / 560


In [9]:
x = train_set
n = 0.1
wrong = 0
right = 0
tp = 0
tn = 0
fp = 0
fn = 0
# Para cada um nos testes
for i in range(len(x)):
    yd = x[i][2]
    xd = np.array([x[i][0],x[i][1]])
    sigma = w[0]*xd[0]+w[1]*xd[1]
    if (sigma >= 0): # função degrau
        y = 1
    else:
        y = 0
    if (y != yd):
        wrong += 1
        if (y == 1):
            fp += 1
        else:
            fn += 1
    else:
        right += 1
        if (y == 1):
            tp += 1
        else:
            tn += 1
confusion_matrix = np.array([[tp,fp],[fn, tn]])
print("Quantidade de certos:",right,"/",len(x))
tpTrain = tp
fpTrain = fp
fnTrain = fn
tnTrain = tn
rightTrain = right
confusion_matrix

Quantidade de certos: 490 / 560


array([[ 56,   0],
       [ 70, 434]])

In [10]:
x = test_set
n = 0.1
wrong = 0
right = 0
tp = 0
tn = 0
fp = 0
fn = 0
# Para cada um nos testes
for i in range(len(x)):
    yd = x[i][2]
    xd = np.array([x[i][0],x[i][1]])
    sigma = w[0]*xd[0]+w[1]*xd[1]
    if (sigma >= 0): # função degrau
        y = 1
    else:
        y = 0
    if (y != yd):
        wrong += 1
        if (y == 1):
            fp += 1
        else:
            fn += 1
    else:
        right += 1
        if (y == 1):
            tp += 1
        else:
            tn += 1

confusion_matrix = np.array([[tp,fp],[fn, tn]])
tpTest = tp
fpTest = fp
fnTest = fn
tnTest = tn
print("Quantidade de certos:",right,"/",len(x))
confusion_matrix

Quantidade de certos: 204 / 240


array([[ 38,   0],
       [ 36, 166]])

## Métricas de desempenho

In [11]:
# Qual a acurácia da solução proposta para os dados do conjunto de treinamento inicialmente fornecido?
acTrain = rightTrain/train_size
print("Acurácia do treino:",acTrain)
# Nos mesmos termos da questão anterior, obtenha os valores de precisão, revocação e F -Score;
precTrain = tpTrain/(tpTrain+fpTrain)
revoTrain = tpTrain/(tpTrain+fnTrain)
FScoreTrain = (2*(precTrain*revoTrain))/(precTrain+revoTrain)
print("Precisão do treino:",precTrain)
print("Revocação do treino:",revoTrain)
print("F-Score do treino:",FScoreTrain)

Acurácia do treino: 0.875
Precisão do treino: 1.0
Revocação do treino: 0.4444444444444444
F-Score do treino: 0.6153846153846153


## A partir destas métricas, discorra acerca da qualidade desta solução perante o conjunto de testes.

Apesar de uma acurácia e precisão boas, esta solução, no entanto, não é muito desejada. Quanto aos negativos, dos verdadeiramente negativos, não errou nenhum. Mas a solução tem precisão de 100% e revocação de 44%, isto indica que quando a solução diz que é positivo, é realmente positivo, mas ele tem muita dificuldade em identificar os verdadeiramente positivos, apesar de reconhecê-los quando os vê.

O F-Score da solução não ajuda muito, tendo 0.615 de pontuação, o que não é muito desejável. Com as métricas em mãos, nossa equipe tomou a conclusão de que esta solução não é de boa qualidade perante o conjunto de testes.